File: backend/app/schemas/document.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class FieldOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    field_key: str
    field_label: str
    predicted_value: str | None
    corrected_value: str | None
    final_value: str | None
    confidence: float
    page: int
    bbox_x: float | None
    bbox_y: float | None
    bbox_w: float | None
    bbox_h: float | None
    is_validated: bool
    is_corrected: bool
    is_position_corrected: bool


class FieldUpdate(BaseModel):
    corrected_value: str | None = None
    is_validated: bool | None = None

    # Manuelle Positionskorrektur: wird gesetzt, wenn der Nutzer im Viewer ein
    # Rechteck über die tatsächliche Fundstelle zieht (z.B. weil das Modell das
    # Feld gar nicht erkannt hatte). Alle vier Bbox-Werte müssen zusammen
    # mitgeschickt werden, `page` ist optional (Default: aktuelle Seite bleibt
    # bzw. wird auf 1 gesetzt, falls noch keine Seite bekannt war).
    page: int | None = None
    bbox_x: float | None = Field(default=None, ge=0, le=1)
    bbox_y: float | None = Field(default=None, ge=0, le=1)
    bbox_w: float | None = Field(default=None, gt=0, le=1)
    bbox_h: float | None = Field(default=None, gt=0, le=1)


class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    owner_id: str
    owner_email: str | None
    filename: str
    content_type: str
    status: str
    page_count: int
    uploaded_at: datetime
    reviewed_at: datetime | None
    error_message: str | None


class DocumentDetailOut(DocumentOut):
    fields: list[FieldOut]


class TrainingExportRow(BaseModel):
    document_id: str
    filename: str
    field_key: str
    field_label: str
    predicted_value: str | None
    final_value: str | None
    was_corrected: bool
    confidence: float
    page: int
    bbox_x: float | None
    bbox_y: float | None
    bbox_w: float | None
    bbox_h: float | None
    was_position_corrected: bool

File: frontend/src/types/document.ts
export interface ContractField {
  id: string;
  field_key: string;
  field_label: string;
  predicted_value: string | null;
  corrected_value: string | null;
  final_value: string | null;
  confidence: number;
  page: number;
  bbox_x: number | null;
  bbox_y: number | null;
  bbox_w: number | null;
  bbox_h: number | null;
  is_validated: boolean;
  is_corrected: boolean;
  is_position_corrected: boolean;
}

export type DocumentStatus =
  | "uploaded"
  | "processing"
  | "needs_review"
  | "reviewed"
  | "failed";

export interface DocumentSummary {
  id: string;
  owner_id: string;
  owner_email: string | null;
  filename: string;
  content_type: string;
  status: DocumentStatus;
  page_count: number;
  uploaded_at: string;
  reviewed_at: string | null;
  error_message: string | null;
}

export interface DocumentDetail extends DocumentSummary {
  fields: ContractField[];
}

export interface TemplateField {
  id: string;
  field_key: string;
  field_label: string;
  sort_order: number;
  patterns: string[] | null;
}

export interface ContractTemplate {
  id: string;
  key: string;
  name: string;
  fields: TemplateField[];
}
